Duplicate from dashboard and from settings

develohpanda opened the pull request:
### Features

* [prevent duplicate if blank name](https://www.loom.com/share/b2e1b75315ad4a6e85bfe09449d55131)
* [collection local to remote from settings (should auto enable sync)](https://www.loom.com/share/b1ccb9cf438848d0ac12707caa7266a8)
* [collection local to remote from dashboard (should auto enable sync and open)](https://www.loom.com/share/f4309aec42c34a41bd56c3de1e5197d0)
* [design local to remote (should **not** enable sync)](https://www.loom.com/share/01141fec550b49b2adc352790c96bd4b)
* collection remote to local (shouldn't enable sync, but should open)
* automatically open the duplicated document to the right activity

<img width="274" alt="image" src="https://user-images.githubusercontent.com/4312346/126743513-49036626-138e-4630-bb72-3f13dcbbad1e.png">

No designs exist for this at the moment, so this is a first pass. 

Introduced react-hook-form to do this; it's a much nicer way to manage form state.

Closes INS-755, Closes INS-721, Closes INS-756

develohpanda committed:
duplicate from dashboard and from settings

develohpanda committed:
use ForwardRefRenderFunction

develohpanda committed:
rename to baseSpace

develohpanda committed:
validation

wdawson reviewed (approved):
Assuming duplicating from a remote to local/base also works ⛵

develohpanda committed:
Merge branch 'develop' of github.com:Kong/insomnia into feature/ins-755-collections-can-be-moved-from-one-remote

develohpanda committed:
Fix conflicts

develohpanda committed:
Merge branch 'develop' into feature/ins-755-collections-can-be-moved-from-one-remote

DMarby commented:
@develohpanda Do we need to wire up any logic here to run the same "Setup sync" logic when duplicating into a Remote space, as we do when you create a new collection in a remote space?

develohpanda committed:
Merge branch 'develop' into feature/ins-755-collections-can-be-moved-from-one-remote

develohpanda committed:
conditionally initialize and push

develohpanda committed:
initialize and mark for sync on duplicate

develohpanda committed:
don't clear the base space when clearing models

develohpanda committed:
activate the correct workspace after duplication

develohpanda committed:
typo

develohpanda committed:
take into account the cached activity

develohpanda committed:
add tests for activating a workspace

develohpanda commented:
@DMarby & @wdawson please see the demos in the description again - I updated this to PR to activate/open the workspace after it is duplicated, regardless of if it's duplicated from the dashboard or from settings.

In addition, I updated this PR such that if a user is duplicating _into_ a remote space, and they are logged in, and they are duplicating a request collection (the only thing supported by Insomnia sync right now), the duplicated workspace will automatically connect with sync.

develohpanda commented on packages/insomnia-app/app/ui/redux/modules/workspace.ts:
```diff
@@ -70,3 +71,31 @@ export const createWorkspace = ({ scope, onCreate }: {
   };
 };
+
+export const activateWorkspace = (workspace: Workspace) => {
```
We activate the workspace in various places; ideally we start using this action in the future, so we correctly activate the space, workspace, and activity

dimitropoulos replied:
If you're aware of some I would thoroughly support you notating them briefly in a ticket so we can remember to do this later.

develohpanda replied:
Drafted a follow-up PR: https://github.com/Kong/insomnia/pull/3856

dimitropoulos replied:
wow! great!  I was just thinkin of having a bulleted list but this is way better! thanks!

dimitropoulos commented on packages/insomnia-app/app/ui/components/modals/workspace-duplicate-modal.tsx:
```diff
@@ -0,0 +1,157 @@
+
+const SpaceOption: FC<Space> = space => (
+  <option key={space._id} value={space._id}>
+    {space.name} ({isBaseSpace(space) ? 'Base' : isLocalSpace(space) ? 'Local' : 'Remote'})
```
since we're using it already in a few places do you think it's worthwhile to put `Base`, `Local`, and `Remote` in `strings.ts` somewhere or something?

develohpanda replied:
That seems like a reasonable thing to do

dimitropoulos commented on packages/insomnia-app/app/ui/components/modals/workspace-duplicate-modal.tsx:
```diff
@@ -0,0 +1,157 @@
+
+const SpaceOption: FC<Space> = space => (
+  <option key={space._id} value={space._id}>
+    {space.name} ({isBaseSpace(space) ? 'Base' : isLocalSpace(space) ? 'Local' : 'Remote'})
```
does... our Dropdown component not work with react-hook-form?

I was gonna 

<img height="150" src="https://user-images.githubusercontent.com/15232461/127239973-1c206304-e402-4860-84f6-da7e959db077.png" />

gets the icons like
<img height="200" src="https://user-images.githubusercontent.com/15232461/127239976-ec88aa00-3147-4fc8-807d-bf0fd6529425.png" />

But I switched it out real quick for `Dropdown` and `DropdownItem` to see and it did not render.

dimitropoulos commented on packages/insomnia-app/app/ui/components/modals/workspace-duplicate-modal.tsx:
```diff
@@ -0,0 +1,157 @@
+}
+
+const SpaceOption: FC<Space> = space => (
+  <option key={space._id} value={space._id}>
```
does... our Dropdown component not work with react-hook-form?

I was gonna 

<img height="150" src="https://user-images.githubusercontent.com/15232461/127239973-1c206304-e402-4860-84f6-da7e959db077.png" />

gets the icons like
<img height="200" src="https://user-images.githubusercontent.com/15232461/127239976-ec88aa00-3147-4fc8-807d-bf0fd6529425.png" />

But I switched it out real quick for `Dropdown` and `DropdownItem` to see and it did not render.

develohpanda replied:
It should work we'll just need to link up some of the events - for this I used a traditional select (none of the forms in Insomnia use the custom dropdown yet)

dimitropoulos replied:
well maybe timebox it to 20 minutes or something but it'd be really great to see how react-hook-form looks/works with our standard dropdown.

develohpanda replied:
Okay! I'll timebox and give it a go, I'd like to use our dropdown as well

dimitropoulos commented on packages/insomnia-app/app/models/helpers/workspace-operations.ts:
```diff
@@ -15,9 +15,10 @@ export async function rename(w: Workspace, s: ApiSpec, name: string) {
 }
 
-export async function duplicate(w: Workspace, name: string) {
+export async function duplicate(w: Workspace, { name, parentId }: Pick<Workspace, 'name' | 'parentId'>) {
```
https://user-images.githubusercontent.com/15232461/127240988-c54f9d72-0a90-4285-8c8e-5bfc3eeb3a52.mp4

develohpanda replied:
This is an existing issue that exists with both the Swagger 2 and OpenAPI 3 importers and the way in which they generate IDs.

In the openapi3 importer, a request id is deterministic and includes the workspace id. In the swagger2 importer, a request id is not yet deterministic, see #3804. The two exhibit similarly broken behavior though.

This is an issue with the importer though, and it needs a bit of an overhaul altogether because we have a lot of id conflict and duplication issues with the way it's implemented at the moment

develohpanda replied:
See [this thread](https://kongstrong.slack.com/archives/GPARXT60G/p1626052892004200?thread_ts=1626048539.003100&cid=GPARXT60G)

develohpanda replied:
We do need to look into this but because it's behavior that already exists in the app I'm inclined to not change in this PR. Now that we're suggesting duplication as an actual user flow to move a workspace between spaces, we need to address the funkiness that arises

dimitropoulos replied:
perfect - just wanted to double triple check that it wasn't a regression in some way or unintended consequence of changes in this PR.  totally not appropriate to solve in this PR - 100% agree

dimitropoulos commented on packages/insomnia-app/app/sync/vcs/initialize-project.ts:
```diff
@@ -39,8 +44,11 @@ export const pushSnapshotOnInitialize = async ({
   const spaceIsForWorkspace = spaceId === workspace.parentId;
   const markedForPush = workspaceMeta?.pushSnapshotOnInitialize;
+  
+  // A race condition causes us to hit this codepath twice while activating a workspace but the first time it has no project so we shouldn't do anything
```
a race between which two things? (just for my own understanding)

develohpanda replied:
Added more details in comments 9d598a6c53e3f9ea0b40b104678a11b65e2ebaeb

dimitropoulos commented on packages/insomnia-app/app/ui/components/modals/workspace-duplicate-modal.tsx:
```diff
@@ -0,0 +1,157 @@
+    });
+
+  const onSubmit = useCallback(async ({ spaceId, newName }: FormFields) => {
+    const duplicateToSpace = spaces.find(s => s._id === spaceId);
```
nothing major but `space => space._id` or `({ _id }) => _id` may be a readability improvement here>

dimitropoulos commented on packages/insomnia-app/app/ui/components/modals/workspace-duplicate-modal.tsx:
```diff
@@ -0,0 +1,157 @@
+        </div>
+        <div className="form-control form-control--outlined">
+          <label>
+            {strings.space.plural} to duplicate into
```
```suggestion
            {strings.space.singular} to duplicate into
```

@wdawson I don't know how much you care about stuff like this so I'm moreso asking to learn your preference, but there's an argument to be made here for this to always be singular because we're always copying into a single space (this isn't a multi-select box at this time)?

<img height="150" src="https://user-images.githubusercontent.com/15232461/127241346-fa2cace2-e689-4ce1-833d-6f73525fd51c.png" />

develohpanda replied:
Oops! Yeah this should probably be singular

develohpanda replied:
[`3d1a59f` (#3820)](https://github.com/Kong/insomnia/pull/3820/commits/3d1a59fdc5e2edaf47b7af4266d0b24bfbf33e7e)

wdawson replied:
Definitely should be singular! Happy to get the pings for this stuff if there's confusion or contention 👍

dimitropoulos commented on packages/insomnia-app/package.json:
```diff
@@ -146,6 +147,7 @@
     "react-dnd": "^7.4.5",
     "react-dnd-html5-backend": "^7.4.4",
     "react-dom": "^16.8.3",
+    "react-hook-form": "^7.11.0",
```
latest is now `7.12.1`

dimitropoulos commented on packages/insomnia-app/app/ui/redux/selectors.ts:
```diff
@@ -378,3 +383,13 @@ export const selectSyncItems = createSelector(
 );
+
+export const selectIsLoggedIn = createSelector(
+  (state: RootState) => state.global,
+  global => global.isLoggedIn,
+);
+
+export const selectActiveActivity = createSelector(
+  (state: RootState) => state.global,
```
should we have `const selectGlobal = (state: RootState) => state.global` since this is duplicated here?

develohpanda replied:
Fixed in 16f7b15fcd1c8d39de05119ad751e82c2b567c9d

dimitropoulos commented on packages/insomnia-app/app/ui/redux/modules/global.tsx:
```diff
@@ -7,7 +7,7 @@ import { combineReducers, Dispatch } from 'redux';
 
 import { trackEvent } from '../../../common/analytics';
-import type { GlobalActivity } from '../../../common/constants';
+import { GlobalActivity } from '../../../common/constants';
```
![200](https://user-images.githubusercontent.com/15232461/127241953-223140ca-3e18-445d-823b-86e97616bbcd.gif)

dimitropoulos commented on packages/insomnia-app/app/ui/redux/modules/__tests__/workspace.test.ts:
```diff
@@ -141,4 +134,138 @@ describe('workspace', () => {
+      await store.dispatch(activateWorkspace(workspace));
+
+      expect(store.getActions()).toEqual([
+        {
+          type: SET_ACTIVE_SPACE,
+          spaceId: space._id,
+        },
+        {
+          type: SET_ACTIVE_WORKSPACE,
+          workspaceId: workspace._id,
+        },
+        {
+          type: SET_ACTIVE_ACTIVITY,
+          activity: ACTIVITY_SPEC,
+        },
```
whenever the time comes, as an aside, redux toolkit is going to greatly reduce the boilerplate here:

```
.toEqual([
  setActiveSpace(space._id),
  setActiveWorkspace(workspace._id),
  setActiveActivity(ACTIVITY_SPEC),
])
```

dimitropoulos commented on packages/insomnia-app/app/ui/redux/modules/__tests__/workspace.test.ts:
```diff
@@ -14,24 +14,15 @@ import { BASE_SPACE_ID } from '../../../../models/space';
 const middlewares = [thunk];
 const mockStore = configureMockStore(middlewares);
 
-const createStoreWithSpace = async () => {
-  const space = await models.initModel(models.space.type);
-
-  const entities = { spaces: { [space._id]: space } };
-  const global = { activeSpaceId: space._id };
-  const store = mockStore({ entities, global });
-  return { store, space };
-};
-
```
wow, nice.

dimitropoulos reviewed (changes requested):
saw some things.  lemme know whatcha think

develohpanda committed:
introduce selectGlobal

develohpanda committed:
convert to singular

develohpanda committed:
extract to strings

develohpanda committed:
readability

develohpanda committed:
update react-hook-form to latest

develohpanda committed:
add more detail to explain race condition

develohpanda committed:
lint fix

dimitropoulos reviewed (approved):
I'd really love to see us use our Dropdown component with react-hook-form, but it's getting to the end of the sprint and it's not worth blocking this if you don't have time.  everything else looks great!

Nice improvements in practically every PR.  What a life we're livin' with this project.  Fun stuff.

develohpanda committed:
Merge branch 'develop' into feature/ins-755-collections-can-be-moved-from-one-remote
